### PR TITLE
Dedup

### PR DESCRIPTION
**Pull request by danamansana**
This addresses SCC-1406, Backend: De-dupe based on itemid. I've set it up to ignore checkouts with ids that have been processed recently, based on a configurable time that I have set to 1 hr. There's not a lot of detail in the ticket, so let me know if you have a different idea as to how we should dedup

### CODE REVIEW

File: lib/item_stream_handler.rb [hunk 1]
```diff
@@ -64,10 +71,13 @@ def handle (event)
         # Presence of 'duedate' indicates it's checked-out
         if decoded && decoded['status'] && ! decoded['status']['duedate'].nil?
           checkout = Checkout.from_item_record decoded
-          add_checkout checkout
-
-          checkout_count += 1
-          update_count checkout
+          unless RECENT_IDS[checkout.id] && RECENT_IDS[checkout.id] - Time.now < ENV[CHECKOUT_ID_EXPIRE_TIME].to_i
```
Comment (nonword) [inline lib/item_stream_handler.rb]: Could we have a test around this functionality? And handle nil ENV var?

Comment (danamansana) [reply lib/item_stream_handler.rb]: I've added some tests. Also, I've made sure to call `.to_i` on all the ENV["CHECKOUT_ID_EXPIRE_TIMES"]'s, so if this is nil it should be equivalent to no deduping. Let me know if you think there is a better default.

File: spec/item_stream_handler_spec.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      end
+
+      it 'should process a checkout with a new id' do
+        item_stream_handler.handle(mock_event)
+        expect(item_stream_handler).to have_received(:add_checkout).once
+      end
+
+      it 'should not process a checkout with a recent id' do
+        item_stream_handler.handle(mock_event)
+        expect(item_stream_handler).not_to have_received(:add_checkout)
```
Comment (nonword) [inline spec/item_stream_handler_spec.rb]: Does this test depend on a side-effect of the first test?

**Review by nonword — CHANGES_REQUESTED** [review]
Test suite isn't passing for me

```
Failures:

  1) ItemStreamHandler#handle deduping based on ids should process a checkout with a new id
     Failure/Error:
       @creation_dates ||= Array.new(checkouts.size)
         .map { |ind| rand delta_seconds(checkouts) }
         .sort
         .reverse
         .map { |s| Time.at(Time.now - s).iso8601 }
     
     NoMethodError:
       undefined method `size' for nil:NilClass
     # ./lib/s3_writer.rb:22:in `creation_dates'
     # ./lib/s3_writer.rb:64:in `feed_xml'
     # ./lib/s3_writer.rb:91:in `write'
     # ./lib/item_stream_handler.rb:87:in `handle'
     # ./spec/item_stream_handler_spec.rb:43:in `block (4 levels) in <top (required)>'

  2) ItemStreamHandler#handle deduping based on ids should process a checkout with an old id
     Failure/Error:
       @creation_dates ||= Array.new(checkouts.size)
         .map { |ind| rand delta_seconds(checkouts) }
         .sort
         .reverse
         .map { |s| Time.at(Time.now - s).iso8601 }
     
     NoMethodError:
       undefined method `size' for nil:NilClass
     # ./lib/s3_writer.rb:22:in `creation_dates'
     # ./lib/s3_writer.rb:64:in `feed_xml'
     # ./lib/s3_writer.rb:91:in `write'
     # ./lib/item_stream_handler.rb:87:in `handle'
     # ./spec/item_stream_handler_spec.rb:54:in `block (4 levels) in <top (required)>'

Finished in 0.03069 seconds (files took 0.45795 seconds to load)
7 examples, 2 failures
```

**Comment by danamansana** [comment]
I've changed the tests so they are no longer relying on side effects to work. They seem to be working now

**Review by nonword — APPROVED** [review]

**Pull request merged**